clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -89,7 +89,7 @@ public FetchResponseData data() {
      */
     public FetchResponse(FetchResponseData fetchResponseData) {
         super(ApiKeys.FETCH);
-        this.data = fetchResponseData;
+        this.data = convertNullRecordsToEmpty(fetchResponseData);
```

Comment:
It would be useful to add a comment that for consistency, we always set records to empty.

Reply:
Make sense 😺

Reply:
This code is called by both the server and the client. On the server side, it makes sense for the server to set records to empty. On the client side, it seems that it's better to take whatever it receives from the broker, instead of changing it silently. Maybe we can consolidate the server side usage based on FetchResponse.of() and the client side usage based on FetchResponse.parse(), and avoid the direct usage of this constructor.

Reply:
Make sense. It could also avoid overheads for clients.
Does this approach also need to apply to `ShareFetchResponse`?

Reply:
Yes, we can use the same approach for `ShareFetchResponse`.

Reply:
https://github.com/apache/kafka/blob/28de78bcbad605a3e906d085d2e59b441ae35212/core/src/main/scala/kafka/server/KafkaApis.scala#L3104-L3111

https://github.com/apache/kafka/blob/28de78bcbad605a3e906d085d2e59b441ae35212/core/src/main/scala/kafka/server/KafkaApis.scala#L3121-L3125

I'm not familiar with the queues feature. Should we add `MemoryRecords.Empty` for these two parts?
cc @AndrewJSchofield 

Many thanks 🙇🏼

Reply:
Those partitions are created using the `ShareFetchResponse.of()` constructor. If we initialize the records in the constructor properly, we don't need to change it here.